File: AndroidTai/app/src/main/java/com/tai/androidtai/dagger/components/DashboardComponent.java
package com.tai.androidtai.dagger.components;

import com.tai.androidtai.dagger.modules.ActivityModule;
import com.tai.androidtai.dagger.modules.DashboardModule;
import com.tai.androidtai.dagger.scope.PerActivity;
import com.tai.androidtai.modules.dashboard.DashboardActivity;

import dagger.Component;

@PerActivity
@Component(dependencies = AndroidTaiComponent.class, modules = {ActivityModule.class, DashboardModule.class})
public interface DashboardComponent {

    final class Initializer {

        private Initializer() {
            throw new UnsupportedOperationException();
        }

        public static DashboardComponent init(AndroidTaiComponent applicationComponent,
                                              ActivityModule activityModule) {
            return DaggerDashboardComponent.builder()
                    .androidTaiComponent(applicationComponent)
                    .activityModule(activityModule)
                    .dashboardModule(new DashboardModule())
                    .build();
        }
    }

    void inject(DashboardActivity activity);

}

File: AndroidTai/app/src/main/java/com/tai/androidtai/dagger/components/DashboardComponent.kt
package com.tai.androidtai.dagger.components

import com.tai.androidtai.dagger.modules.ActivityModule
import com.tai.androidtai.dagger.modules.DashboardModule
import com.tai.androidtai.dagger.scope.PerActivity
import com.tai.androidtai.modules.dashboard.DashboardActivity
import com.tai.androidtai.modules.mealDetails.MealDetailsActivity

import dagger.Component

@PerActivity
@Component(dependencies = [AndroidTaiComponent::class], modules = [ActivityModule::class, DashboardModule::class])
interface DashboardComponent {

    class Initializer private constructor() {

        init {
            throw UnsupportedOperationException()
        }

        companion object {

            fun init(applicationComponent: AndroidTaiComponent,
                     activityModule: ActivityModule): DashboardComponent {
                return DaggerDashboardComponent.builder()
                        .androidTaiComponent(applicationComponent)
                        .activityModule(activityModule)
                        .dashboardModule(DashboardModule())
                        .build()
            }
        }
    }

    fun inject(dashboardActivity: DashboardActivity)

    fun inject(mealDetailsActivity: MealDetailsActivity)
}
